README.md
Magento Full Catalog Translate
==============================

Magento module to translate all your products automatically using Google Translate API.

How does it work
----------------
When you install the module it will create a new product attribute called "Translate automatically?" (you'll find it in the "General" tab for all of your products).

Let's say you have your main store in English, then you create a storeview in Italian, now all of your products in the Italian store view have the English texts. What you want is to quickly have them in Italian. Be sure the locales are correctly configured for all the stores/storeviews you're going to use in the process (this module will automatically read the language settings from the storeviews).

So go to your "Catalog -> Manage products" mask in the Magento backend, select the Italian store view (or anyway the one you want to translate), then with the mass action tools selects all the products and "update attributes", now set the "Translate automatically?" attribute to yes. This means that these products for this store view need to be translated (we'll see later how and from what language to what language).

Now we have products in English and products marked as "to be translated" in our Italian storeview.

Let's step back a little bit: configuration.
Before actually translating our contents you need to review some configurations of the module, go to the Magento backend mask "System -> Configuration -> Services -> Fballiano Full Catalog Translate".
Here can decide which product attributes have to be translated (defaults: name, short_description, description, meta_title, meta_keyword, meta_description) but most importantly you'll have to fill your Google Translate API key (without it nothing will actually work).

Open the console and navigate to the "shell" directory.
Now we'll start the actual translate process.

```shell
php fballiano_full_catalog_translate.php sourcestorecode targetstorecode
```

in out case it could be:
```shell
php fballiano_full_catalog_translate.php default storeviewita
```

The process will gather all the products that need to be translated (from the target storeview), gather the untranslated text (from the source storeview), call Google Translate API for every attribute, import the translated text into the target store view, set the record to "not to be translated" (again, into the target storeview).

Google Translate characters limit!
----------------------------------
Google Translate APIs only supports a maximum of 5,000 characters in a single request. If you setup this module to use Google Translate, be sure that your products' descriptions are less than 5,000 characters long!

Backup!!!
---------
Backup your database before launching the translation process!!!
This module is provided "as is" and I'll not be responsible for any data damage.

Installation
------------

Simply download the whole repository and copy everything to your Magento document root.
Otherwise with modman:
```shell
modman clone https://github.com/fballiano/magento-full-catalog-translate
```

Compatibility
-------------
This module was developed on Magento 1.9.
If you have a different version of Magento and the module is working please drop me a line so I can update this compatibility list.

Some technical detail
---------------------
To speed up a bit the import process, a patched version of [Magmi](http://sourceforge.net/projects/magmi/) it's included in the extension (lib/Magmi).

TODO
----

Remove magmi and migrate to https://github.com/avstudnitz/AvS_FastSimpleImport

Support
-------
If you have any issues with this extension, open an issue on GitHub.

Contribution
------------
Any contributions are highly appreciated. The best way to contribute code is to open a
[pull request on GitHub](https://help.github.com/articles/using-pull-requests).

Developer
---------
Fabrizio Balliano  
[http://fabrizioballiano.com](http://fabrizioballiano.com)  
[@fballiano](https://twitter.com/fballiano)

Licence
-------
[OSL - Open Software Licence 3.0](http://opensource.org/licenses/osl-3.0.php)

Copyright
---------
(c) Fabrizio Balliano


CHANGELOG.md
0.7.3
-----
* code reformat
* Google Translate API URL was changed (now is "https://translation.googleapis.com/language/translate/v2")

0.7.2
-----
* mbstring extension is required (check added) if you use Google Translate API.
* if you translate an attribute that has a text longer than 5000 chars with
  Google Translate API, the script will output a warning and skip that specific
  attribute for that specific product.

0.7.1
-----
* solved issue #12

0.7.0
-----
* categories translation was added

0.6.1
-----
* fixed modman file

0.6.0
-----
* support for debug and dry run was added to the fballiano_full_catalog_translate.php shell script

0.5.0
-----
* support for custom translation command was introduced

0.4.0
-----
* a bug with Magmi and table prefixes was solved

0.3.0
-----
* "Attribute to translate" config il now a multiselect

0.2.0
-----
* Table names read dinamically

0.1.0
-----
* Initial release


shell/fballiano_full_catalog_translate.php
<?php
/**
 * FBalliano
 *
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is bundled with this package in the file LICENSE.txt.
 * It is also available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * DISCLAIMER
 *
 * Do not edit or add to this file if you wish to upgrade this Module to
 * newer versions in the future.
 *
 * @category   FBalliano
 * @package    FBalliano_FullCatalogTranslate
 * @copyright  Copyright (c) 2014 Fabrizio Balliano (http://fabrizioballiano.it)
 * @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
 */

require_once 'abstract.php';

class Fballiano_FullCatalogTranslate_Shell extends Mage_Shell_Abstract
{
    protected $helper = null;
    protected $translation_system = null;
    protected $api_key = null;
    protected $command = null;
    protected $store_source = null;
    protected $store_dest = null;
    protected $language_source = null;
    protected $language_dest = null;
    protected $ws_url = "https://translation.googleapis.com/language/translate/v2";
    protected $attributes_to_translate = null;
    protected $category_attributes_to_translate = null;
    protected $datapump = null;
    protected $debug_mode = false;
    protected $dry_run = false;

    public function run()
    {
        $this->debug_mode = $this->getArg("debug");
        $this->dry_run = $this->getArg("dry");
        if ($this->dry_run) $this->debug_mode = true;
        $this->helper = Mage::helper("fballiano_fullcatalogtranslate");
        $this->translation_system = $this->helper->getTranslationSystem();
        $this->attributes_to_translate = $this->helper->getAttributesToTranslate();
        $this->category_attributes_to_translate = $this->helper->getCategoryAttributesToTranslate();
        $this->api_key = $this->helper->getApiKey();
        $this->ws_url .= "?key={$this->api_key}";
        $this->command = $this->helper->getCommand();

        switch ($this->translation_system) {
            case "googletranslate":
                if (!$this->api_key) die("Please set your API key in the Magento admin configuration.\n");
                if (!function_exists("mb_strlen")) die("Please install mbstring PHP extension\n");
                break;
            case "custom":
                if (!$this->command) die("Please set the translation command in the Magento admin configuration.\n");
                break;
            default:
                die("Unrecognized translation system: {$this->translation_system}.\n");
        }

        $args = array_keys($this->_args);
        $this->store_source = @$args[0];
        $this->store_dest = @$args[1];
        if (!$this->store_source or !$this->store_dest) die($this->usageHelp());

        require_once "Magmi/inc/magmi_defs.php";
        require_once "Magmi/integration/inc/magmi_datapump.php";
        require_once "Magmi/integration/inc/productimport_datapump.php";
        require_once "Magmi/integration/inc/productimport_datapump.php";
        $this->datapump = new Magmi_ProductImport_Datapump();
        $this->datapump->beginImportSession("default", "create");

        $appEmulation = Mage::getSingleton("core/app_emulation");

        try {
            $initialEnvironmentInfo = $appEmulation->startEnvironmentEmulation($this->store_dest);
        } catch (Mage_Core_Model_Store_Exception $e) {
            die("Target store view \"{$this->store_dest}\" doesn't seem to exist\n" . $this->usageHelp());
        }

        $store_id_dest = Mage::app()->getStore()->getId();
        $this->language_dest = substr(Mage::app()->getLocale()->getLocaleCode(), 0, 2);
        $this->ws_url .= "&target={$this->language_dest}";

        $appEmulation->stopEnvironmentEmulation($initialEnvironmentInfo);

        $attribute_id = Mage::getModel("catalog/entity_attribute")->loadByCode(Mage_Catalog_Model_Product::ENTITY, "fb_translate")->getId();
        $table_name = Mage::getSingleton("core/resource")->getTableName("catalog_product_entity_int");
        $products = Mage::getSingleton("core/resource")->getConnection("core_read")->fetchCol("SELECT entity_id FROM {$table_name} WHERE attribute_id={$attribute_id} AND store_id={$store_id_dest} AND value=1");

        try {
            $initialEnvironmentInfo = $appEmulation->startEnvironmentEmulation($this->store_source);
        } catch (Mage_Core_Model_Store_Exception $e) {
            die("Source store \"{$this->store_source}\" view doesn't seem to exist\n" . $this->usageHelp());
        }

        $this->language_source = substr(Mage::app()->getLocale()->getLocaleCode(), 0, 2);
        $this->ws_url .= "&source={$this->language_source}";

        $this->translateCategories($store_id_dest);

        $product = Mage::getModel("catalog/product");
        foreach ($products as $product_id) {
            $product->load($product_id);
            $row = $product->getData();

            echo "Translating {$row["sku"]} from {$this->language_source} to {$this->language_dest}... ";
            if ($this->debug_mode) echo "\n";
            $translated_row = array();
            $translated_row["store"] = (string)$this->store_dest;
            $translated_row["sku"] = (string)$row["sku"];
            $translated_row["fb_translate"] = "0"; //leave it as string otherwise magmi won't save it
            foreach ($this->attributes_to_translate as $attribute) {
                if (strlen($row[$attribute])) {
                    $translated_row[$attribute] = $this->translateString($row[$attribute]);
                }
                if ($this->debug_mode) {
                    echo "\t[$attribute] [{$row[$attribute]}] -> [{$translated_row[$attribute]}]\n";
                }
            }
            if (!$this->dry_run) $this->datapump->ingest($translated_row);
            echo "OK\n";
        }

        $appEmulation->stopEnvironmentEmulation($initialEnvironmentInfo);
        $this->datapump->endImportSession();
        echo "Terminated\n";
    }

    public function translateCategories($store_id_dest)
    {
        $appEmulation = Mage::getSingleton("core/app_emulation");
        $attribute_id = Mage::getModel("catalog/entity_attribute")->loadByCode(Mage_Catalog_Model_Category::ENTITY, "fb_translate")->getId();
        $table_name = Mage::getSingleton("core/resource")->getTableName("catalog_category_entity_int");
        $categories = Mage::getSingleton("core/resource")->getConnection("core_read")->fetchCol("SELECT entity_id FROM {$table_name} WHERE attribute_id={$attribute_id} AND store_id={$store_id_dest} AND value=1");

        foreach ($categories as $category_id) {
            $initialEnvironmentInfo = $appEmulation->startEnvironmentEmulation($this->store_source);
            $category = Mage::getModel("catalog/category");
            $category->load($category_id);
            $row = $category->getData();
            $appEmulation->stopEnvironmentEmulation($initialEnvironmentInfo);

            echo "Translating category {$row["entity_id"]} from {$this->language_source} to {$this->language_dest}... ";
            if ($this->debug_mode) echo "\n";
            $translated_row = array();
            $translated_row["fb_translate"] = 0;
            foreach ($this->category_attributes_to_translate as $attribute) {
                if (strlen($row[$attribute])) {
                    $translated_row[$attribute] = $this->translateString($row[$attribute]);
                }
                if ($this->debug_mode) {
                    echo "\t[$attribute] [{$row[$attribute]}] -> [{$translated_row[$attribute]}]\n";
                }
            }
            if (!$this->dry_run) {
                $initialEnvironmentInfo = $appEmulation->startEnvironmentEmulation($this->store_dest);
                $category = Mage::getModel("catalog/category");
                $category->load($category_id);
                $category->addData($translated_row);
                $category->save();
                $appEmulation->stopEnvironmentEmulation($initialEnvironmentInfo);
            }
            echo "OK\n";
        }
    }

    public function productCollectionWalkCallback($args)
    {
        $row = $args["row"];
        echo "Translating {$row["sku"]} from {$this->language_source} to {$this->language_dest}... ";
        if ($this->debug_mode) echo "\n";
        $translated_row = array();
        $translated_row["store"] = $this->store_dest;
        $translated_row["sku"] = $row["sku"];
        $translated_row["fb_translate"] = 0;
        foreach ($this->attributes_to_translate as $attribute) {
            if (strlen($row[$attribute])) {
                if ($this->translation_system == "googletranslate" and mb_strlen($translated_row[$attribute]) >= 5000) {
                    echo "\t[$attribute] more than 5000 chars long, unsupported by Google Translate, not translated\n";
                    continue;
                }
                $translated_row[$attribute] = $this->translateString($row[$attribute]);
                if ($this->debug_mode) {
                    echo "\t[$attribute] [{$row[$attribute]}] -> [{$this->language_dest} {$translated_row[$attribute]}]\n";
                }
            }
        }

        if (!$this->dry_run) $this->datapump->ingest($translated_row);
        echo "OK\n";
    }

    public function translateString($string)
    {
        switch ($this->translation_system) {
            case "googletranslate":
                $ws_url = "{$this->ws_url}&q=" . urlencode($string);
                if ($this->debug_mode) echo "\t{$ws_url}\n";
                $translated = json_decode(file_get_contents($ws_url), true);
                return (string)$translated["data"]["translations"][0]["translatedText"];
            case "custom":
                $command = str_replace(
                    array("%SOURCELANGUAGE%", "%TARGETLANGUAGE%", "%STRING%"),
                    array($this->language_source, $this->language_dest, $string),
                    $this->command
                );
                if ($this->debug_mode) echo "\t{$command}\n";
                return shell_exec($command);
        }
    }

    public function usageHelp()
    {
        return <<<USAGE

Usage:\tphp -f fballiano_full_catalog_translate.php source_store_view_code target_store_view_code [-debug] [-dry]

-debug\tshows every call to traslation systems and the results
-dry\ttranslates everything (automatically enables debug also) but do not save anything on the database

USAGE;
    }
}

$shell = new Fballiano_FullCatalogTranslate_Shell();
$shell->run();